File: opensource/kernel/src/main/java/com/sos/scheduler/engine/kernel/job/Task.java
package com.sos.scheduler.engine.kernel.job;

import com.sos.scheduler.engine.cplusplus.runtime.Sister;
import com.sos.scheduler.engine.cplusplus.runtime.SisterType;
import com.sos.scheduler.engine.data.job.TaskId;
import com.sos.scheduler.engine.eventbus.EventSource;
import com.sos.scheduler.engine.kernel.cppproxy.TaskC;
import com.sos.scheduler.engine.kernel.order.UnmodifiableOrder;

import javax.annotation.Nullable;

public final class Task implements UnmodifiableTask, Sister, EventSource {
    private final TaskC cppProxy;

    protected Task(TaskC cppProxy) {
        this.cppProxy = cppProxy;
    }

    @Override public void onCppProxyInvalidated() {}

    public static class Type implements SisterType<Task, TaskC> {
        @Override public final Task sister(TaskC proxy, Sister context) {
            return new Task(proxy);
        }
    }

    @Override public Job getJob() {
        return cppProxy.job().getSister();
    }

    @Override public String toString() {
        return Task.class.getSimpleName();
    }

	@Override @Nullable public UnmodifiableOrder getOrderOrNull() {
		return cppProxy.order().getSister();
	}

	@Override public TaskId getId() {
		return new TaskId( cppProxy.id() );
	}
}

File: opensource/kernel/src/main/scala/com/sos/scheduler/engine/kernel/job/Task.scala
package com.sos.scheduler.engine.kernel.job

import com.sos.scheduler.engine.cplusplus.runtime.Sister
import com.sos.scheduler.engine.data.job.TaskId
import com.sos.scheduler.engine.eventbus.EventSource
import com.sos.scheduler.engine.kernel.cppproxy.TaskC
import com.sos.scheduler.engine.kernel.order.UnmodifiableOrder
import java.io.File
import javax.annotation.Nullable

final class Task(cppProxy: TaskC) extends UnmodifiableTask with Sister with EventSource {

  def onCppProxyInvalidated() {}

  def job =
    cppProxy.job.getSister

  @Nullable def getOrderOrNull: UnmodifiableOrder =
    cppProxy.order.getSister

  def id =
    new TaskId(cppProxy.id)

  def parameterValue(name: String): String =
    cppProxy.params.get_string(name)

  def stdoutFile =
    new File(cppProxy.stdout_path)

  def stderrFile =
    new File(cppProxy.stderr_path)

  override def toString =
    s"Task($id)"
}
